text/044-vo-named-format.md

```diff
@@ -0,0 +1,163 @@
+end of the file that contains information about the segments, and in particular
+their names. Merely parsing this metadata is enough to know the position of
+a segment inside the file. The position of this registry is known by parsing
+the header of the file.
+
+Another design choice is the reliance on 64-bit integers to indicate file
+positions, in order to support large files. The current COFF was probably
+already problematic for some files, e.g. some `vio` files can easily reach the
+4GB range which would overflow positions.
+
+A minor change introduced *en passant* is the separation between the magic
+number and the version number. The rationale is that many tools rely on a fixed
+magic number to classify files. Coq uses the version number as a magic number,
+which means that it is harder to quickly check that a file is indeed a Coq
+object file. In the new COFF, we have a fixed magic number followed by the
+version number.
+
+## Formal description
+
+All integers use the big-endian representation. We will describe objects as
+lists of subcomponents, given by their type, length and semantics.
+
+```
+-- object file --
+int32   | 4 bytes  | magic number (0x43 0x6F 0x71 0x21) "Coq!"
+int32   | 4 bytes  | Coq version
+int64   | 8 bytes  | absolute position of the summary
+seg₁    \
+...     | variable | list of segments
+segₙ    /
+summary | variable | summary
+
+-- segment --
+...     | variable | binary data
+hash    | 16 bytes | MD5 of the above data
+
+-- summary --
+int32   | 4 bytes  | number of segments (n below)
+sum₁    \
+...     | variable | segment summaries
+sumₙ    /
+
+-- segment summary --
+string  | variable | name of the segment
+int64   | 8 bytes  | absolute position of the segment
+int64   | 8 bytes  | segment length (without hash)
+hash    | 16 bytes | MD5 (same as the one in the segment)
+
+-- string --
+int32   | 4 bytes  | length of the next field
+data    | variable | ASCII string
+```
+
+We use a 32-bit integer for the number of segments, but it looks unlikely we
+will ever need that many segments.
+
+Note that the actual data stored in a segment is not specified. We happen to
+use OCaml marshalled data, but there is no reason for this. Since we use names,
+we can actually store arbitrary data there and adopt a different serialization
+scheme on a name-based mechanism. For instance, VM bytecode could be stored
+as a raw string in a dedicated segment.
+
+# Drawbacks
+
+The new COFF is a bit more verbose, since it needs to store additional metadata.
+In practice this metadata should be small compared to the actual data stored in
+the files. Thus, it should not affect file size much.
+
+# Alternatives
+
+We could have used a standard file format instead of rolling up our own.
+Unfortunately I did not find a format that was satisfying all the following
+constraints.
+- The container could store uncompressed data
+- The data is random-access
+- The format does not suffer easy-to-hit limitations.
```

Comment:
What about key/value store implementations such as Berkeley DB?

Reply:
:+1: I think Dune uses a variation of that and it works pretty well indeed